subxt/src/events.rs

```diff
@@ -38,17 +37,360 @@ use codec::{
+    }
+
+    /// Iterate over all of the events, using metadata to dynamically
+    /// decode them as we go, and returning the raw bytes and other associated
+    /// details.
+    ///
+    /// This method is safe to use even if you do not statically know about
+    /// all of the possible events; it splits events up using the metadata
+    /// obtained at runtime, which does.
+    pub fn iter_raw(
+        &self,
+    ) -> impl Iterator<Item = Result<RawEventDetails, BasicError>> + '_ {
+        let event_bytes = &self.event_bytes;
+
+        let mut pos = 0;
+        let mut index = 0;
+        std::iter::from_fn(move || {
+            let cursor = &mut &event_bytes[pos..];
+            let start_len = cursor.len();
+
+            if start_len == 0 || self.num_events == index {
+                None
+            } else {
+                match decode_raw_event_details::<T>(self.metadata, index, cursor) {
+                    Ok(raw_event) => {
+                        // Skip over decoded bytes in next iteration:
+                        pos += start_len - cursor.len();
+                        // Increment the index:
+                        index += 1;
+                        // Return the event details:
+                        Some(Ok(raw_event))
+                    }
+                    Err(e) => {
+                        // By setting the position to the "end" of the event bytes,
+                        // the cursor len will become 0 and the iterator will return `None`
+                        // from now on:
+                        pos = event_bytes.len();
+                        Some(Err(e))
+                    }
+                }
+            }
+        })
+    }
+
+    /// Iterate through the events using metadata to dynamically decode and skip
+    /// them, and return only those which should decode to the provided `Ev` type.
+    pub fn find<Ev: Event>(&self) -> impl Iterator<Item = Result<Ev, BasicError>> + '_ {
+        self.iter_raw().filter_map(|ev| {
+            ev.and_then(|ev| ev.as_event::<Ev>().map_err(Into::into))
+                .transpose()
+        })
+    }
+
+    /// Iterate through the events using metadata to dynamically decode and skip
+    /// them, and return the first event found which decodes to the provided `Ev` type.
+    pub fn find_first_event<Ev: Event>(&self) -> Result<Option<Ev>, BasicError> {
+        self.find::<Ev>().next().transpose()
+    }
+
+    /// Find an event that decodes to the type provided. Returns true if it was found.
+    pub fn has<Ev: crate::Event>(&self) -> Result<bool, BasicError> {
+        Ok(self.find::<Ev>().next().transpose()?.is_some())
+    }
+}
+
+/// A decoded event and associated details.
+#[derive(Debug, Clone, PartialEq)]
+pub struct EventDetails<Evs> {
+    /// When was the event produced?
+    pub phase: Phase,
+    /// What index is this event in the stored events for this block.
+    pub index: u32,
+    /// The event itself.
+    pub event: Evs,
```

Comment:
My OCD tickles a bit on the "s" in "Evs". Can it be just "Ev"?

Reply:
I found uses of `Ev` (see `find`, `find_first_event`, `has`), so I'll keep the `s` to help differentiate even if it is a bit ew :D